src/main/java/tdl/client/queue/QueueBasedImplementationRunner.java

```diff
@@ -44,78 +220,214 @@ public Builder withSolutionFor(String methodName, UserImplementation userImpleme
             deployProcessingRules
                     .on(methodName)
                     .call(userImplementation)
-                    .then(publish());
-            return this;
-        }
-
-        public Builder withSolutionFor(String methodName, UserImplementation userImplementation, ClientAction action) {
-            deployProcessingRules
-                    .on(methodName)
-                    .call(userImplementation)
-                    .then(action);
+                    .build();
             return this;
         }
 
         public QueueBasedImplementationRunner create() {
+            logToConsole("        QueueBasedImplementationRunner.Builder create");
+            logToConsole("        QueueBasedImplementationRunner.Builder deployProcessingRules: " + deployProcessingRules);
             return new QueueBasedImplementationRunner(config, deployProcessingRules);
         }
 
         private static ProcessingRules createDeployProcessingRules() {
+            logToConsole("        QueueBasedImplementationRunner.Builder createDeployProcessingRules [start]");
             ProcessingRules deployProcessingRules = new ProcessingRules();
 
             // Debt - we only need this to consume message from the server
             deployProcessingRules
                     .on("display_description")
                     .call(params -> "OK")
-                    .then(publish());
+                    .build();
 
+            logToConsole("        QueueBasedImplementationRunner.Builder createDeployProcessingRules [end]");
             return deployProcessingRules;
         }
     }
 
     public void run() {
+        logToConsole("        QueueBasedImplementationRunner run [start]");
         audit.logLine("Starting client");
-        try (RemoteBroker remoteBroker = new RemoteBroker(config.getHostname(), config.getPort(), config.getUniqueId(), config.getRequestTimeoutMillis())){
-            //Design: We use a while loop instead of an ActiveMQ MessageListener to process the messages in order
+        requests.clear();
+        try {
+            requests.clear();
+            requests.addAll(receive());
```

Comment:
This is fetching all the requests before we start to process them. Not what we have discussed.
We should not request for a new message until we have processed the current ones.
We should drop the buffering of the receive() requests.

The `deployProcessingRules.getResponseFor(request);` should happen right in the middle of receive `while` loop.

Reply:
Ah, yes it must have slipped through, I can now apply this refactoring as the others are now out of the way.